src/session.ts

```diff
@@ -0,0 +1,248 @@
+      clientName: 'JupyterLab',
+      adapterID: 'python',
+      pathFormat: 'path',
+      linesStartAt1: true,
+      columnsStartAt1: true,
+      supportsVariableType: true,
+      supportsVariablePaging: true,
+      supportsRunInTerminalRequest: true,
+      locale: 'en-us'
+    });
+
+    await this.sendRequest('attach', {});
+  }
+
+  /**
+   * Stop the running debug session.
+   */
+  public async stop(): Promise<void> {
+    await this.sendRequest('disconnect', {
+      restart: false,
+      terminateDebuggee: true
+    });
+  }
+
+  /**
+   * Send a custom debug request to the kernel.
+   * @param command debug command.
+   * @param args arguments for the debug command.
+   */
+  async sendRequest<K extends keyof DebugSession.IDebugRequestTypes>(
+    command: K,
+    args: DebugSession.IDebugRequestTypes[K]
+  ): Promise<DebugSession.IDebugResponseTypes[K]> {
+    const message = await this._sendDebugMessage({
+      type: 'request',
+      seq: this._seq++,
+      command,
+      arguments: args
+    });
+    return message.content as DebugSession.IDebugResponseTypes[K];
+  }
+
+  /**
+   * Signal emitted for debug event messages.
+   */
+  get eventMessage(): ISignal<DebugSession, DebugProtocol.Event> {
+    return this._eventMessage;
+  }
+
+  /**
+   * Handle debug events sent on the 'iopub' channel.
+   */
+  private _handleEvent(
+    sender: IClientSession,
+    message: KernelMessage.IIOPubMessage
+  ): void {
+    const msgType = message.header.msg_type;
+    if (msgType !== 'debug_event') {
+      return;
+    }
+    const event = message.content as DebugProtocol.Event;
+    this._eventMessage.emit(event);
+  }
+
+  /**
+   * Send a debug request message to the kernel.
+   * @param msg debug request message to send to the kernel.
+   */
+  private async _sendDebugMessage(
+    msg: KernelMessage.IDebugRequestMsg['content']
+  ): Promise<KernelMessage.IDebugReplyMsg> {
+    const reply = new PromiseDelegate<KernelMessage.IDebugReplyMsg>();
+    const kernel = this.client.kernel;
+    const future = kernel.requestDebug(msg);
+    future.onReply = (msg: KernelMessage.IDebugReplyMsg) => {
+      if (!msg.content.success) {
+        return reply.reject(msg);
+      }
+      return reply.resolve(msg);
+    };
+    await future.done;
+    return reply.promise;
+  }
+
+  client: IClientSession;
+  editors: CodeEditor.IEditor[];
+
+  private _disposed = new Signal<this, void>(this);
+  private _isDisposed: boolean = false;
+
+  private _eventMessage = new Signal<DebugSession, DebugProtocol.Event>(this);
+
+  /**
+   * Debug protocol sequence number.
+   */
+  private _seq: number = 0;
+}
+
+/**
+ * A namespace for `DebugSession` statics.
+ */
+export namespace DebugSession {
+  export interface IOptions {
+    /**
+     * The client session used by the debug session.
+     */
+    client: IClientSession;
+  }
+
+  /**
+   * Interface for all the debug requests types.
+   */
+  export interface IDebugRequestTypes {
```

Comment:
How do you feel about changing this from `DebugSession.IDebugRequestTypes` to `IDebugger.Requests`?

So that you could use it in this way:
```typescript
export const attach = async (
  args: IDebugger.Requests['attach']
): Promise<void> => Promise.resolve(undefined);
```

Reply:
From @jtpio on Gitter:
> don't we want Requests and Responses to be for a session rather than a debugger?

So you are proposing this?

```typescript
const attach = async (
  args: IDebugger.ISession.Requests['attach']
): Promise<IDebugger.ISession.Responses['attach']> => {
  return { body: null } as IDebugger.ISession.Responses['attach'];
};
```

Reply:
Yes that would be the idea

Reply:
I admit it's a bit "nested" though

Reply:
Yeah, I agree, but I think it does make sense to do it this way even if it is deeper in the namespace.

Reply:
This actually becomes `IDebugger.ISession.IResponses` (leading "I" to follow the tslint rule)

Reply:
Instead of `IResponses` I was considering `type Responses` and `type Requests`. What are your thoughts?

Reply:
yes sure, that works.